File: java_bindings/lib/src/main/java/arpirobot/core/network/NetworkTable.java
/*
 * Copyright 2021 Marcus Behel
 *
 * This file is part of ArPiRobot-CoreLib.
 * 
 * ArPiRobot-CoreLib is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * ArPiRobot-CoreLib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with ArPiRobot-CoreLib.  If not, see <https://www.gnu.org/licenses/>. 
 */

package arpirobot.core.network;

import arpirobot.Bridge;
import com.sun.jna.Pointer;


/**
 * Helper class used to manage network table key/value pairs
 */
public class NetworkTable {
    /**
     * Sets a given key/value pair. If the key does not exist it will be created. Else the value will be updated.
     * @param key The key for the pair
     * @param value The value for the pair
     */
    public static void set(String key, String value){
        Bridge.arpirobot.NetworkTable_set(key, value);
    }

    /**
     * Get the value for a key/value pair
     * @param key The key to get the associated value with
     * @return The value associated with the given key. If the key does not exist an empty string is returned.
     */
    public static String get(String key){
        Pointer ptr = Bridge.arpirobot.NetworkTable_get(key);
        String str = ptr.getString(0);
        Bridge.arpirobot.freeString(ptr);
        return str;
    }

    /**
     * Check if a key has a value
     * @param key The key to check for a value associated with
     * @return true if a value exists for the given key, else false
     */
    public static boolean has(String key){
        return Bridge.arpirobot.NetworkTable_has(key);
    }
}

File: cpp_library/src/arpirobot/core/network/NetworkTable.cpp
/*
 * Copyright 2021 Marcus Behel
 *
 * This file is part of ArPiRobot-CoreLib.
 * 
 * ArPiRobot-CoreLib is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * ArPiRobot-CoreLib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with ArPiRobot-CoreLib.  If not, see <https://www.gnu.org/licenses/>. 
 */

#include <arpirobot/core/network/NetworkTable.hpp>
#include <arpirobot/core/network/NetworkManager.hpp>
#include <arpirobot/core/log/Logger.hpp>
#include <arpirobot/core/robot/BaseRobot.hpp>
#include <cmath>
#include <sstream>
#include <iomanip>
#include <algorithm>


using namespace arpirobot;
using namespace std::placeholders;


std::unordered_map<std::string, std::string> NetworkTable::data;
std::mutex NetworkTable::lock;
bool NetworkTable::inSync = false;

void NetworkTable::set(std::string key, std::string value){
    setFromRobot(key, value);
}

std::string NetworkTable::get(std::string key){
    std::lock_guard<std::mutex> l(lock);
    auto it = data.find(key);
    if(it != data.end()){
        return data[key];
    }
    return "";
}

bool NetworkTable::has(std::string key){
    std::lock_guard<std::mutex> l(lock);
    auto it = data.find(key);
    return it != data.end();
}


bool NetworkTable::isInSync(){
    return inSync;
}

void NetworkTable::startSync(){
    lock.lock();
    inSync = true;
    Logger::logDebug("Starting sync from robot to DS.");

    if(!NetworkManager::sendNtRaw(asio::buffer(NET_TABLE_START_SYNC_DATA, 3))){
        // abortSync will be called by handleDisconnect and lock will be released
        return;
    }

    sendAllValues();
    if(!inSync){
        return;
    }

    Logger::logDebug("Ending sync from robot to DS. Waiting for DS to sync data to robot.");
    NetworkManager::sendNtRaw(asio::buffer(NET_TABLE_END_SYNC_DATA, 4));
}

void NetworkTable::sendAllValues(){
    for(const auto &it : data){
        if(!inSync){
            return;
        }
        NetworkManager::sendNt(it.first, it.second);
    }
}

void NetworkTable::finishSync(std::unordered_map<std::string, std::string> dataFromDs){
    Logger::logDebug("Got all sync data from DS to robot.");

    for(const auto &it : dataFromDs){
        data[it.first] = it.second;
    }

    inSync = false;
    lock.unlock();
    Logger::logInfo("Network table sync complete.");
}

void NetworkTable::abortSync(){
    inSync = false;
    lock.unlock();
    Logger::logWarning("Network table sync aborted.");
}

void NetworkTable::setFromRobot(std::string key, std::string value){
    // Do not allow \n or 255 in the key or value
    std::replace(key.begin(), key.end(), (char)255, '\0');
    std::replace(key.begin(), key.end(), '\n', '\0');
    {
        std::lock_guard<std::mutex> l(lock);
        data[key] = value;
    }
    NetworkManager::sendNt(key, value);
}

void NetworkTable::setFromDs(std::string key, std::string value){
    std::lock_guard<std::mutex> l(lock);
    data[key] = value;
}
